compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -142,7 +142,8 @@ trait TypeAssigner {
     val mbr = qualType.findMember(name, pre)
     if (reallyExists(mbr))
       qualType.select(name, mbr)
-    else if (qualType.derivesFrom(defn.DynamicClass) && name.isTermName && !Dynamic.isDynamicMethod(name))
+    else if (qualType.derivesFrom(defn.DynamicClass) && name.isTermName && !Dynamic.isDynamicMethod(name) &&
+        !(name == nme.apply && tree.span.isSynthetic && Dynamic.isDynamicMethod(qual1.symbol.name)))
```

Comment:
The following code compiles before, but fails to compile after the changes:

```Scala
import scala.language.dynamics

class Foo extends Dynamic {
  def applyDynamic(name: String): Bar = ???
}

class Bar extends Dynamic {
  def applyDynamic(name: String)(x: Int): Qux = ???
}

class Qux extends Dynamic {
  def applyDynamic(name: String)(x: Int) = ???
}


var foo = new Foo
def baz = foo.blah(42).blah(45)
```

Reply:
That is intended, as in Scala 2.
https://scastie.scala-lang.org/gsxJ9jifRQyc5jLTbrmYSQ

Reply:
The only worry I have is about using positions for semantics (though we have precedent for that when dealing with contextual functions), as positions are very weak invariants.

If there are no better alternatives, maybe we add the example as a neg test?

Reply:
Indeed. I will try to find another way to identify the application

Reply:
Now I handle properly contextual functions, but we still need to check the positions.